src/sign.rs

```diff
@@ -0,0 +1,269 @@
+/*
+ *  Copyright (c) 2020 Samsung Electronics Co., Ltd All Rights Reserved
+ *
+ *  Contact: Lukasz Pawelczyk <l.pawelczyk@samsung.com>
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License
+ */
+
+use libc::{c_void, c_char};
+use std::ptr;
+
+use crate::yaca_common as common;
+use crate::yaca_lib as lib;
+use crate::yaca_conv as conv;
+use crate::crypto::{Context, ContextWithPadding};
+use crate::*;
+
+
+/// Context for `Sign` operations
+pub struct SignContext {
+    handle: *mut c_void,
+}
+
+impl Drop for SignContext {
+    fn drop(&mut self)
+    {
+        unsafe {
+            lib::yaca_context_destroy(self.handle)
+        }
+    }
+}
+
+impl Context for SignContext {
+    fn get_handle(&self) -> *mut c_void
+    {
+        self.handle
+    }
+}
+
+impl ContextWithPadding for SignContext {}
+
+impl SignContext {
+    /// Initializes a signature context for asymmetric signatures
+    ///
+    /// - `algo` is a digest algorithm that will be used.
+    /// - `prv_key` is a private key that will be used, algorithm is
+    ///   deduced based on key type, supported key types:
```

Comment:
These are effectively checked at runtime deep down under FFI. That's ok, it mirrors the original API and keeps the number of distinct entities minimal.

However, introducing distinct enums is not without merit either as It would prevent such errors at compile time.

By no means am I insisting on a change here, just dumping my later Friday's brain.

Reply:
You need to elaborate here. What new distinct enums you talk about introducing here?

Reply:
In hindsight I shouldn't have posted this comment. It's unclear and belongs in a religious discussion rather than this PR. Resolving.

Now the elaboration.
Let's have the following pseudocode, before:
```
enum Wuj: Wujek | Wujaszek | Wujcio  // three variants
fn accept_any(Wuj w);
fn refuse_wujcio(Wuj w); // compiles but panics when fed Wujcio
```
After:
```
enum Wuj: Wujek | Wujaszek | Wujcio  // three variants
fn accept_any(Wuj w);

enum WujButNotWujcio: Wujek | Wujaszek; // two variants
fn refuse_wujcio(WujButNotWujcio w);
```

Reply:
No, I get the idea of creating separate enums. I was asking what enums specifically you meant in this particular case? You marked this comment on a line that describes the private key. And I don't really see any area of improvement here.

Reply:
Ah. I don't have enough domain knowledge to make a constructive suggestion. The only thing that comes to mind is to prevent all cases of enums permitting invalid values, methodically, one by one.

I commented on this line because it reminded me of the "issue" and I didn't feel like scrolling to an appropriate line, for whatever reason ;)

Reply:
Right.
Anyway, I get the point, but still I'd want this API to be relatively close to the C API. I won't be able to eliminate all EINVALs anyway so wouldn't want to overly complicate this.

I don't want however to have any panics when it's the user's error (in regards to your `uncle` code). Panics (debug asserts) are only in case either yaca or yaca-rust fails. If the user fails he should get EINVAL (or some other API compatible error). So if you such case please take note.